src/NuGet.Clients/NuGet.VisualStudio/SolutionRestoreManager/IVsTargetFrameworkInfo4.cs

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Collections.Generic;
+
+namespace NuGet.SolutionRestoreManager
+{
+    /// <summary>
+    /// A type representing all project inputs (items and properties) for the given target framework.
+    /// It is implemented by project systems sending nominations to NuGet.
+    /// </summary>
+    public interface IVsTargetFrameworkInfo4
+    {
+        /// <summary>
+        /// Collection of item types. The dictionary key is the MSBuild item type (for example, PackageReference), and
+        /// the value is the list of items of that item type.
+        /// </summary>
+        IReadOnlyDictionary<string, IReadOnlyList<IVsReferenceItem2>>? Items { get; }
```

Comment:
I am sure you may have thought about it, but I am curious if we are losing any type-safety benefits by using a bag of properties populated by the project system while nominating a project for restore. I understand the limitations of the existing design where we have to add new interfaces every time, we want the project system to pass a new MSBuild property during nomination, as it requires changes on the project system side as well.

Reply:
All our other restore paths are already use similar code, because MSBuild's APIs are similar to this PR's generic item bag. For example, see:

* https://github.com/NuGet/NuGet.Client/pull/5679/files#diff-69bcc4cc18dd020ef093e419c357cd2a3452094ec27be239a03c665b32de863b
* https://github.com/NuGet/NuGet.Client/pull/5714/files#diff-1c1b03f2151db1b5a3e33e52b3eb6421948e0cc949a4333acb21e960e9b2f06c

There might be more risk by using a generic bag, rather than properties with specific names, but that was just offloading the responsibility to dotnet/project-system. It doesn't avoid some code somewhere needing to pick out items by name out of a generic API.

It may warrant closer scrutiny in pull requests, but it's similar code to what we're doing for non-SDK projects in VS, and also all projects on the command line.